runtime/gc_base/StringTable.cpp

```diff
@@ -976,4 +975,53 @@ j9gc_internString(J9VMThread *vmThread, j9object_t sourceString)
 	return internedString;
 }
 
+typedef struct {
+	J9UTF8* utf;
+	j9object_t stringObject;
+} J9UTFCacheEntry;
+
+static UDATA
+utfCacheHashFn(void *key, void *userData)
+{
+	/* UTF8 pointers are 2-aligned, so shift the pointer for maximum entropy */
+	return ((UDATA)((J9UTFCacheEntry*)key)->utf) >> 1;
+}
+
+static UDATA
+utfCacheHashEqualFn(void *leftKey, void *rightKey, void *userData)
+{
+	return ((J9UTFCacheEntry*)leftKey)->utf == ((J9UTFCacheEntry*)rightKey)->utf;
+}
+
+j9object_t
+j9gc_createJavaLangStringWithUTFCache(J9VMThread *vmThread, J9UTF8 *utf)
+{
+	J9HashTable *utfCache = vmThread->utfCache;
+	j9object_t result = NULL;
+
+	if (NULL == utfCache) {
+		J9JavaVM *vm = vmThread->javaVM;
+		utfCache = hashTableNew(OMRPORT_FROM_J9PORT(vm->portLibrary), J9_GET_CALLSITE(), 0, sizeof(J9UTFCacheEntry), sizeof(void*), 0, J9MEM_CATEGORY_VM, utfCacheHashFn, utfCacheHashEqualFn, NULL, NULL);
+		/* Failure to allocate the cache is non-fatal and is ignored */
+		if (NULL != utfCache) {
+			vmThread->utfCache = utfCache;
+		}
+	} else {
+		J9UTFCacheEntry exemplar = { utf, NULL };
+		J9UTFCacheEntry *entry = (J9UTFCacheEntry*)hashTableFind(utfCache, &exemplar);
+		if (NULL != entry) {
+			result = J9WEAKROOT_OBJECT_LOAD(vmThread, &entry->stringObject);
+			goto done;
+		}
+	}
+	result = j9gc_createJavaLangString(vmThread, J9UTF8_DATA(utf), J9UTF8_LENGTH(utf), J9_STR_INTERN);
+	if ((NULL != utfCache) && (NULL != result)) {
```

Comment:
Is there a use-after free bug here?  Consider the following flow:
* read `utfCache` from the thread, 
* don't find the string in the cache, 
* allocate which causes a GC and frees the `utfCache`
* come back to this code and see a non-null `utfCache` pointing at a released hashtable

Reply:
Good catch - will update.